File: ios/oplayer/Model/ModelUtils.m
//
//  ModelUtils.m
//  oplayer
//
//  Created by SYALON on 13-9-11.
//
//

#import "ModelUtils.h"

#import "ChainObjectManager.h"
#import "OrgUtils.h"
#import "TradingPair.h"

@implementation ModelUtils

/*
 *  (public) 资产 - 判断资产是否允许强清
 */
+ (BOOL)assetCanForceSettle:(id)asset_object
{
    NSInteger flags = [[[asset_object objectForKey:@"options"] objectForKey:@"flags"] integerValue];
    if ((flags & ebat_disable_force_settle) != 0) {
        return NO;
    }
    return YES;
}

/*
 *  (public) 资产 - 判断资产是否允许发行人全局清算
 */
+ (BOOL)assetCanGlobalSettle:(id)asset_object
{
    NSInteger issuer_permissions = [[[asset_object objectForKey:@"options"] objectForKey:@"issuer_permissions"] integerValue];
    return (issuer_permissions & ebat_global_settle) != 0;
}

/*
 *  (public) 资产 - 判断资产是否允许隐私转账
 */
+ (BOOL)assetAllowConfidential:(id)asset_object
{
    NSInteger flags = [[[asset_object objectForKey:@"options"] objectForKey:@"flags"] integerValue];
    if ((flags & ebat_disable_confidential) != 0) {
        return NO;
    }
    return YES;
}

/*
 *  (public) 资产 - 资产是否允许覆盖转账（强制转账）
 */
+ (BOOL)assetCanOverride:(id)asset_object
{
    NSInteger flags = [[[asset_object objectForKey:@"options"] objectForKey:@"flags"] integerValue];
    return (flags & ebat_override_authority) != 0;
}

/*
 *  (public) 资产 - 是否所有转账都需要发行人审核
 */
+ (BOOL)assetIsTransferRestricted:(id)asset_object
{
    NSInteger flags = [[[asset_object objectForKey:@"options"] objectForKey:@"flags"] integerValue];
    return (flags & ebat_transfer_restricted) != 0;
}

/*
 *  (public) 资产 - 资产是否需要持有人属于白名单判断。
 */
+ (BOOL)assetNeedWhiteList:(id)asset_object
{
    NSInteger flags = [[[asset_object objectForKey:@"options"] objectForKey:@"flags"] integerValue];
    return (flags & ebat_white_list) != 0;
}

/*
 *  (public) 资产 - 是否已经全局清算判断
 */
+ (BOOL)assetHasGlobalSettle:(id)bitasset_object
{
    return ![self isNullPrice:[bitasset_object objectForKey:@"settlement_price"]];
}

/*
 *  (public) 资产 - 是否是智能币判断
 */
+ (BOOL)assetIsSmart:(id)asset
{
    NSString* bitasset_data_id = [asset objectForKey:@"bitasset_data_id"];
    return bitasset_data_id && ![bitasset_data_id isEqualToString:@""];
}

/*
 *  (public) 资产 - 是否是链核心资产判断
 */
+ (BOOL)assetIsCore:(id)asset
{
    return [[asset objectForKey:@"id"] isEqualToString:[ChainObjectManager sharedChainObjectManager].grapheneCoreAssetID];
}

/*
 *  (public) 判断是否价格无效
 */
+ (BOOL)isNullPrice:(id)price
{
    NSString* core_asset_id = [ChainObjectManager sharedChainObjectManager].grapheneCoreAssetID;
    if ([core_asset_id isEqualToString:[[price objectForKey:@"base"] objectForKey:@"asset_id"]] &&
        [core_asset_id isEqualToString:[[price objectForKey:@"quote"] objectForKey:@"asset_id"]]) {
        return YES;
    }
    return NO;
}

/*
 *  (public) 辅助方法 - 从full account data获取指定资产等余额信息，返回 NSDecimalNumber 对象，没有找到对应资产则返回 ZERO 对象。
 */
+ (NSDecimalNumber*)findAssetBalance:(NSDictionary*)full_account_data asset_id:(NSString*)asset_id asset_precision:(NSInteger)asset_precision
{
    assert(full_account_data);
    assert(asset_id);
    id balances = [full_account_data objectForKey:@"balances"];
    if (balances) {
        for (id balance_object in balances) {
            if ([asset_id isEqualToString:[balance_object objectForKey:@"asset_type"]]) {
                return [NSDecimalNumber decimalNumberWithMantissa:[[balance_object objectForKey:@"balance"] unsignedLongLongValue]
                                                         exponent:-asset_precision
                                                       isNegative:NO];
            }
        }
    }
    return [NSDecimalNumber zero];
}

+ (NSDecimalNumber*)findAssetBalance:(NSDictionary*)full_account_data asset:(NSDictionary*)asset
{
    assert(asset);
    return [self findAssetBalance:full_account_data
                         asset_id:[asset objectForKey:@"id"]
                  asset_precision:[[asset objectForKey:@"precision"] integerValue]];
}

/*
 *  (public) 从石墨烯ID列表获取依赖的ID列表。
 */
+ (NSArray*)collectDependence:(NSArray*)source_oid_list level_keys:(id)keystring_or_keyarray
{
    if ([keystring_or_keyarray isKindOfClass:[NSString class]]) {
        keystring_or_keyarray = @[keystring_or_keyarray];
    }
    ChainObjectManager* chainMgr = [ChainObjectManager sharedChainObjectManager];
    NSMutableDictionary* id_hash = [NSMutableDictionary dictionary];
    for (id oid in source_oid_list) {
        id obj = [chainMgr getChainObjectByID:oid];
        assert(obj);
        id target_obj = obj;
        for (id level_key in keystring_or_keyarray) {
            assert([target_obj isKindOfClass:[NSDictionary class]]);
            target_obj = [target_obj objectForKey:level_key];
            if (!target_obj) {
                break;
            }
        }
        if (target_obj && [target_obj isKindOfClass:[NSString class]]) {
            [id_hash setObject:@YES forKey:target_obj];;
        }
    }
    return [id_hash allKeys];
}

/*
 *  (public) 计算平均数
 */
+ (NSDecimalNumber*)calculateAverage:(NSDecimalNumber*)total n:(NSDecimalNumber*)n result_precision:(NSInteger)result_precision
{
    NSDecimalNumberHandler* handler = [NSDecimalNumberHandler decimalNumberHandlerWithRoundingMode:NSRoundDown
                                                                                             scale:result_precision
                                                                                  raiseOnExactness:NO
                                                                                   raiseOnOverflow:NO
                                                                                  raiseOnUnderflow:NO
                                                                               raiseOnDivideByZero:NO];
    
    return [total decimalNumberByDividingBy:n withBehavior:handler];
}

/*
 *  (public) 计算总数
 */
+ (NSDecimalNumber*)calTotal:(NSDecimalNumber*)avg n:(NSDecimalNumber*)n result_precision:(NSInteger)result_precision
{
    NSDecimalNumberHandler* handler = [NSDecimalNumberHandler decimalNumberHandlerWithRoundingMode:NSRoundDown
                                                                                             scale:result_precision
                                                                                  raiseOnExactness:NO
                                                                                   raiseOnOverflow:NO
                                                                                  raiseOnUnderflow:NO
                                                                               raiseOnDivideByZero:NO];
    
    return [avg decimalNumberByMultiplyingBy:n withBehavior:handler];
}

/*
 *  (public) 处理链上返回的限价单信息，方便UI显示。
 *  filterTradingPair - 筛选当前交易对相关订单，可为nil。
 */
+ (NSMutableArray*)processLimitOrders:(NSArray*)limit_orders filter:(TradingPair*)filterTradingPair
{
    NSMutableArray* dataArray = [NSMutableArray array];
    if (!limit_orders) {
        return dataArray;
    }
    
    ChainObjectManager* chainMgr = [ChainObjectManager sharedChainObjectManager];
    
    for (id order in limit_orders) {
        id sell_price = [order objectForKey:@"sell_price"];
        id base = [sell_price objectForKey:@"base"];
        id quote = [sell_price objectForKey:@"quote"];
        id base_id = [base objectForKey:@"asset_id"];
        id quote_id = [quote objectForKey:@"asset_id"];
        
        //  筛选当前交易对相关订单，并根据当前交易对确定买卖方向。
        BOOL issell;
        if (filterTradingPair) {
            if ([base_id isEqualToString:filterTradingPair.baseId] && [quote_id isEqualToString:filterTradingPair.quoteId]){
                //  买单：卖出 CNY
                issell = NO;
            }else if ([base_id isEqualToString:filterTradingPair.quoteId] && [quote_id isEqualToString:filterTradingPair.baseId]){
                //  卖单：卖出 BTS
                issell = YES;
            }else{
                //  其他交易对的订单
                continue;
            }
        }
        
        id base_asset = [chainMgr getChainObjectByID:base_id];
        id quote_asset = [chainMgr getChainObjectByID:quote_id];
        assert(base_asset);
        assert(quote_asset);
        
        NSInteger base_precision = [[base_asset objectForKey:@"precision"] integerValue];
        NSInteger quote_precision = [[quote_asset objectForKey:@"precision"] integerValue];
        double base_value = [OrgUtils calcAssetRealPrice:base[@"amount"] precision:base_precision];
        double quote_value = [OrgUtils calcAssetRealPrice:quote[@"amount"] precision:quote_precision];
        
        //  REMARK：没筛选的情况下，根据资产优先级自动计算买卖方向。
        if (!filterTradingPair) {
            id assetBasePriority = [chainMgr genAssetBasePriorityHash];
            NSInteger base_priority = [[assetBasePriority objectForKey:[base_asset objectForKey:@"symbol"]] integerValue];
            NSInteger quote_priority = [[assetBasePriority objectForKey:[quote_asset objectForKey:@"symbol"]] integerValue];
            if (base_priority > quote_priority) {
                issell = NO;
            } else {
                issell = YES;
            }
        }
        
        double price;
        NSString* price_str;
        NSString* amount_str;
        NSString* total_str;
        NSString* base_sym;
        NSString* quote_sym;
        //  REMARK: base 是卖出的资产，除以 base 则为卖价(每1个 base 资产的价格)。反正 base / quote 则为买入价。
        if (!issell){
            //  buy     price = base / quote
            price = base_value / quote_value;
            price_str = [OrgUtils formatFloatValue:price precision:base_precision];
            double total_real = [OrgUtils calcAssetRealPrice:order[@"for_sale"] precision:base_precision];
            double amount_real = total_real / price;
            amount_str = [OrgUtils formatFloatValue:amount_real precision:quote_precision];
            total_str = [OrgUtils formatAssetString:order[@"for_sale"] precision:base_precision];
            base_sym = [base_asset objectForKey:@"symbol"];
            quote_sym = [quote_asset objectForKey:@"symbol"];
        }else{
            //  sell    price = quote / base
            price = quote_value / base_value;
            price_str = [OrgUtils formatFloatValue:price precision:quote_precision];
            amount_str = [OrgUtils formatAssetString:order[@"for_sale"] precision:base_precision];
            double for_sale_real = [OrgUtils calcAssetRealPrice:order[@"for_sale"] precision:base_precision];
            double total_real = price * for_sale_real;
            total_str = [OrgUtils formatFloatValue:total_real precision:quote_precision];
            base_sym = [quote_asset objectForKey:@"symbol"];
            quote_sym = [base_asset objectForKey:@"symbol"];
        }
        //  REMARK：特殊处理，如果按照 base or quote 的精度格式化出价格为0了，则扩大精度重新格式化。
        if ([price_str isEqualToString:@"0"]){
            price_str = [OrgUtils formatFloatValue:price precision:8];
        }
        
        [dataArray addObject:@{@"time":order[@"expiration"],
                               @"issell":@(issell),
                               @"price":price_str,
                               @"amount":amount_str,
                               @"total":total_str,
                               @"base_symbol":base_sym,
                               @"quote_symbol":quote_sym,
                               @"id": order[@"id"],
                               @"seller": order[@"seller"],
                               @"raw_order": order  //  原始数据
        }];
    }
    
    //  按照ID降序排列
    [dataArray sortUsingComparator:(^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
        NSDecimalNumber* n1 = [NSDecimalNumber decimalNumberWithString:[[[obj1 objectForKey:@"id"] componentsSeparatedByString:@"."] lastObject]];
        NSDecimalNumber* n2 = [NSDecimalNumber decimalNumberWithString:[[[obj2 objectForKey:@"id"] componentsSeparatedByString:@"."] lastObject]];
        return [n2 compare:n1];
    })];
    return dataArray;
}

@end

File: android/app/src/main/java/com/btsplusplus/fowallet/utils/ModelUtils.kt
package com.btsplusplus.fowallet.utils

import bitshares.*
import com.btsplusplus.fowallet.kline.TradingPair
import com.fowallet.walletcore.bts.ChainObjectManager
import org.json.JSONArray
import org.json.JSONObject
import java.math.BigDecimal

class ModelUtils {

    companion object {


        /**
         *  (public) 资产 - 判断资产是否允许强清
         */
        fun assetCanForceSettle(asset_object: JSONObject): Boolean {
            val flags = asset_object.getJSONObject("options").getInt("flags")
            if (flags.and(EBitsharesAssetFlags.ebat_disable_force_settle.value) != 0) {
                return false
            }
            return true
        }

        /**
         *  (public) 资产 - 判断资产是否允许发行人全局清算
         */
        fun assetCanGlobalSettle(asset_object: JSONObject): Boolean {
            val issuer_permissions = asset_object.getJSONObject("options").getInt("issuer_permissions")
            return issuer_permissions.and(EBitsharesAssetFlags.ebat_global_settle.value) != 0
        }

        /**
         *  (public) 资产 - 判断资产是否允许隐私转账
         */
        fun assetAllowConfidential(asset_object: JSONObject): Boolean {
            val flags = asset_object.getJSONObject("options").getInt("flags")
            if (flags.and(EBitsharesAssetFlags.ebat_disable_confidential.value) != 0) {
                return false
            }
            return true
        }

        /**
         *  (public) 资产 - 资产是否允许覆盖转账（强制转账）
         */
        fun assetCanOverride(asset_object: JSONObject): Boolean {
            val flags = asset_object.getJSONObject("options").getInt("flags")
            return flags.and(EBitsharesAssetFlags.ebat_override_authority.value) != 0
        }

        /**
         *  (public) 资产 - 是否所有转账都需要发行人审核
         */
        fun assetIsTransferRestricted(asset_object: JSONObject): Boolean {
            val flags = asset_object.getJSONObject("options").getInt("flags")
            return flags.and(EBitsharesAssetFlags.ebat_transfer_restricted.value) != 0
        }

        /**
         *  (public) 资产 - 资产是否需要持有人属于白名单判断。
         */
        fun assetNeedWhiteList(asset_object: JSONObject): Boolean {
            val flags = asset_object.getJSONObject("options").getInt("flags")
            return flags.and(EBitsharesAssetFlags.ebat_white_list.value) != 0
        }

        /**
         *  (public) 资产 - 是否已经全局清算判断
         */
        fun assetHasGlobalSettle(bitasset_object: JSONObject): Boolean {
            return !isNullPrice(bitasset_object.getJSONObject("settlement_price"))
        }

        /**
         *  (public) 资产 - 是否是智能币判断
         */
        fun assetIsSmart(asset: JSONObject): Boolean {
            val bitasset_data_id = asset.optString("bitasset_data_id", "")
            return bitasset_data_id.isNotEmpty()
        }

        /**
         *  (public) 资产 - 是否是链核心资产判断
         */
        fun assetIsCore(asset: JSONObject): Boolean {
            return asset.getString("id") == ChainObjectManager.sharedChainObjectManager().grapheneCoreAssetID
        }

        /**
         *  (public) 判断是否价格无效
         */
        fun isNullPrice(price: JSONObject): Boolean {
            val core_asset_id = ChainObjectManager.sharedChainObjectManager().grapheneCoreAssetID
            if (core_asset_id == price.getJSONObject("base").getString("asset_id") &&
                    core_asset_id == price.getJSONObject("quote").getString("asset_id")) {
                return true
            }
            return false
        }

        /**
         *  (public) 根据汇率换算手续费。
         */
        fun multiplyAndRoundupNetworkFee(core_asset: JSONObject, asset: JSONObject, n_core_fee: BigDecimal, core_exchange_rate: JSONObject): BigDecimal? {
            val core_asset_id = core_asset.getString("id")
            val core_precision = core_asset.getInt("precision")

            val asset_id = asset.getString("id")
            val precision = asset.getInt("precision")

            val n_zero = BigDecimal.ZERO

            val base = core_exchange_rate.getJSONObject("base")
            val quote = core_exchange_rate.getJSONObject("quote")
            if (core_asset_id == base.getString("asset_id")) {
                if (asset_id != quote.getString("asset_id")) {
                    //  null price
                    return null
                }
                //  quote / base * n_core_fee
                val n_base = bigDecimalfromAmount(base.getString("amount"), core_precision)
                if (n_base <= n_zero) {
                    //  error price
                    return null
                }
                val n_quote = bigDecimalfromAmount(quote.getString("amount"), precision)
                return n_core_fee.multiply(n_quote).divide(n_base, precision, BigDecimal.ROUND_UP)
            } else if (core_asset_id == quote.getString("asset_id")) {
                if (asset_id != base.getString("asset_id")) {
                    //  null price
                    return null
                }
                //  base / quote * n_core_fee
                val n_quote = bigDecimalfromAmount(quote.getString("amount"), core_precision)
                if (n_quote <= n_zero) {
                    //  error price
                    return null
                }
                val n_base = bigDecimalfromAmount(base.getString("amount"), precision)
                return n_core_fee.multiply(n_base).divide(n_quote, precision, BigDecimal.ROUND_UP)
            } else {
                assert(false) { "invalid core_exchange_rate" }
            }
            return null
        }

        /**
         *  (public) 辅助方法 - 从full account data获取指定资产等余额信息，返回 NSDecimalNumber 对象，没有找到对应资产则返回 ZERO 对象。
         */
        fun findAssetBalance(full_account_data: JSONObject, asset_id: String, asset_precision: Int): BigDecimal {
            val balances = full_account_data.optJSONArray("balances")
            if (balances != null && balances.length() > 0) {
                for (balance_object in balances.forin<JSONObject>()) {
                    if (asset_id == balance_object!!.getString("asset_type")) {
                        return bigDecimalfromAmount(balance_object.getString("balance"), asset_precision)
                    }
                }
            }
            return BigDecimal.ZERO
        }

        fun findAssetBalance(full_account_data: JSONObject, asset: JSONObject): BigDecimal {
            return findAssetBalance(full_account_data, asset.getString("id"), asset.getInt("precision"))
        }

        /**
         *  (public) 从石墨烯ID列表获取依赖的ID列表。
         */
        fun collectDependence(source_oid_list: JSONArray, keystring_or_keyarray: Any): JSONArray {
            val ary: JSONArray
            if (keystring_or_keyarray is String) {
                ary = jsonArrayfrom(keystring_or_keyarray)
            } else {
                ary = keystring_or_keyarray as JSONArray
            }
            val chainMgr = ChainObjectManager.sharedChainObjectManager()
            val id_hash = JSONObject()
            for (oid in source_oid_list.forin<String>()) {
                var target_obj: Any?
                val obj = chainMgr.getChainObjectByID(oid!!)
                target_obj = obj
                for (level_key in ary.forin<String>()) {
                    val target_obj_as_hash = target_obj as JSONObject
                    target_obj = target_obj_as_hash.opt(level_key!!)
                    if (target_obj == null) {
                        break
                    }
                }
                if (target_obj != null && target_obj is String) {
                    id_hash.put(target_obj, true)
                }
            }
            return id_hash.keys().toJSONArray()
        }

        /**
         *  (public) 计算平均数
         */
        fun calculateAverage(total: BigDecimal, n: BigDecimal, result_precision: Int): BigDecimal {
            return total.divide(n, result_precision, BigDecimal.ROUND_DOWN)
        }

        /**
         *  (public) 计算总数
         */
        fun calTotal(avg: BigDecimal, n: BigDecimal, result_precision: Int): BigDecimal {
            return avg.multiply(n).setScale(result_precision, BigDecimal.ROUND_DOWN)
        }

        /**
         *  (public) 处理链上返回的限价单信息，方便UI显示。
         *  filterTradingPair - 筛选当前交易对相关订单，可为nil。
         */
        fun processLimitOrders(limit_orders: JSONArray?, filterTradingPair: TradingPair?): MutableList<JSONObject> {
            val dataArray = mutableListOf<JSONObject>()

            if (limit_orders == null) {
                return dataArray
            }

            val chainMgr = ChainObjectManager.sharedChainObjectManager()
            for (order in limit_orders.forin<JSONObject>()) {
                val sell_price = order!!.getJSONObject("sell_price")
                val base = sell_price.getJSONObject("base")
                val quote = sell_price.getJSONObject("quote")
                val base_id = base.getString("asset_id")
                val quote_id = quote.getString("asset_id")

                //  筛选当前交易对相关订单，并根据当前交易对确定买卖方向。
                var issell = false
                if (filterTradingPair != null) {
                    if (base_id == filterTradingPair._baseId && quote_id == filterTradingPair._quoteId) {
                        //  买单：卖出 CNY
                        issell = false
                    } else if (base_id == filterTradingPair._quoteId && quote_id == filterTradingPair._baseId) {
                        //  卖单：卖出 BTS
                        issell = true
                    } else {
                        //  其他交易对的订单
                        continue
                    }
                }
                val base_asset = chainMgr.getChainObjectByID(base_id)
                val quote_asset = chainMgr.getChainObjectByID(quote_id)
                val base_precision = base_asset.getInt("precision")
                val quote_precision = quote_asset.getInt("precision")
                val base_value = OrgUtils.calcAssetRealPrice(base.getString("amount"), base_precision)
                val quote_value = OrgUtils.calcAssetRealPrice(quote.getString("amount"), quote_precision)
                //  REMARK：没筛选的情况下，根据资产优先级自动计算买卖方向。
                if (filterTradingPair == null) {
                    val assetBasePriority = chainMgr.genAssetBasePriorityHash()
                    val base_priority = assetBasePriority.optInt(base_asset.getString("symbol"), 0)
                    val quote_priority = assetBasePriority.optInt(quote_asset.getString("symbol"), 0)
                    issell = base_priority <= quote_priority
                }

                //  REMARK: base 是卖出的资产，除以 base 则为卖价(每1个 base 资产的价格)。反正 base / quote 则为买入价。
                var price: Double
                var price_str: String
                var amount_str: String
                var total_str: String
                var base_sym: String
                var quote_sym: String
                if (!issell) {
                    //  buy     price = base / quote
                    price = base_value / quote_value
                    price_str = OrgUtils.formatFloatValue(price, base_precision)
                    val total_real = OrgUtils.calcAssetRealPrice(order.getString("for_sale"), base_precision)
                    val amount_real = total_real / price
                    amount_str = OrgUtils.formatFloatValue(amount_real, quote_precision)
                    total_str = OrgUtils.formatAssetString(order.getString("for_sale"), base_precision)
                    base_sym = base_asset.getString("symbol")
                    quote_sym = quote_asset.getString("symbol")
                } else {
                    //  sell    price = quote / base
                    price = quote_value / base_value
                    price_str = OrgUtils.formatFloatValue(price, quote_precision)
                    amount_str = OrgUtils.formatAssetString(order.getString("for_sale"), base_precision)
                    val for_sale_real = OrgUtils.calcAssetRealPrice(order.getString("for_sale"), base_precision)
                    val total_real = price * for_sale_real
                    total_str = OrgUtils.formatFloatValue(total_real, quote_precision)
                    base_sym = quote_asset.getString("symbol")
                    quote_sym = base_asset.getString("symbol")
                }
                //  REMARK：特殊处理，如果按照 base or quote 的精度格式化出价格为0了，则扩大精度重新格式化。
                if (price_str == "0") {
                    price_str = OrgUtils.formatFloatValue(price, 8)
                }

                dataArray.add(JSONObject().apply {
                    put("time", order.getString("expiration"))
                    put("issell", issell)
                    put("price", price_str)
                    put("amount", amount_str)
                    put("total", total_str)
                    put("base_symbol", base_sym)
                    put("quote_symbol", quote_sym)
                    put("id", order.getString("id"))
                    put("seller", order.getString("seller"))
                    put("raw_order", order)
                })
            }
            //  按照ID降序
            dataArray.sortByDescending { it.getString("id").split(".").last().toInt() }
            //  返回
            return dataArray
        }
    }
}
